[Test] Creating TC_CADMIN_1_22 and TC_CADMIN_1_24 python test modules

j-ororke opened the pull request:
- Created CADMIN_1_22 python test module following steps in PR here: [TC_CADMIN_1_22](https://github.com/CHIP-Specifications/chip-test-plans/pull/4673)
- Created CADMIN_1_24 python test module following steps in PR here: [TC_CADMIN_1_24](https://github.com/CHIP-Specifications/chip-test-plans/pull/4674)
- Removed following yaml test modules: 
- - Test_TC_CADMIN_1_21.yaml, Test_TC_CADMIN_1_22.yaml, Test_TC_CADMIN_1_23.yaml, Test_TC_CADMIN_1_24.yaml

j-ororke committed:
[Test] Creating TC_CADMIN_1_22 and TC_CADMIN_1_24:
- Created CADMIN_1_22 python test module following steps in PR here:
- Created CADMIN_1_24 python test module following steps in PR here:
- Removed following yaml test modules:
-- Test_TC_CADMIN_1_21.yaml, Test_TC_CADMIN_1_22.yaml, Test_TC_CADMIN_1_23.yaml, Test_TC_CADMIN_1_24.yaml

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updating TC_CADMIN_1_22 and TC_CADMIN_1_24:
- Resolved linting issues

cecille commented on src/python_testing/TC_CADMIN_1_22.py:
```diff
@@ -0,0 +1,140 @@
+            logging.info('Commissioning complete done. Successful? {}, errorcode = {}'.format(
+                errcode.is_success, errcode))
+            asserts.assert_false(errcode.is_success, 'Commissioning complete did not error as expected')
+            asserts.assert_true(errcode.sdk_code == 0x0000000B,
```
is there a codegen enum value you could use here? It seems like these values should appear in the enums section of this cluster

j-ororke replied:
Sorry this function that is mentioned above is not in these tests, this code for that function has been removed in this test.
However, I agree that the enum value for the errcode.sdk_code should have an associated value with its cluster, I will find it and replace it in the other tests such as CADMIN_1_3.

cecille commented on src/python_testing/TC_CADMIN_1_24.py:
```diff
@@ -0,0 +1,140 @@
+#
+#    Copyright (c) 2024 Project CHIP Authors
```
There look basically the same with just differences in the success / failure times, right? Can you put these in the same file (you can have more than one test_ function in a file), and just parameterize the test function?

j-ororke replied:
Thank you, i did not realize I could do that.
I have removed the separate test modules and created one merged test module for both CADMIN_1_22 and CADMIN_1_24 tests.

cecille commented on src/python_testing/TC_CADMIN_1_22.py:
```diff
@@ -0,0 +1,140 @@
+
+        self.step(4)
+        window_status = await self.get_window_status()
+        if window_status != 0:
```
is there an enum that indicates 0 as closed?

j-ororke replied:
There actually was!!
I have found it (Clusters.AdministratorCommissioning.Enums.CommissioningWindowStatusEnum.kWindowNotOpen) and replaced it in associated spots in the test module.

j-ororke commented on src/python_testing/TC_CADMIN_1_22.py:
```diff
@@ -0,0 +1,140 @@
+            logging.exception('Error running OpenCommissioningWindow %s', e)
+            asserts.assert_true(False, 'Failed to open commissioning window')
+
+    def generate_unique_random_value(self, value):
```
Sorry, this is not used currently, I have removed it from this test

j-ororke committed:
Created TC_CADMIN_1_22_24 test module:
- Merged TC_CADMIN_1_24 and TC_CADMIN_1_22 standalone test modules into a single test module
- Found enum for window closed value and replaced it in the tests
- Removed generate_unique_value and AttemptCommission functions
- Updated CI arguments format to YAML'esque format

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updated TC_CADMIN_1_22_24:
- Renamed class to TC_CADMIN_1_22_24

j-ororke committed:
Updating TC_CADMIN_1_22_24:
- Resolving lint errors

j-ororke committed:
Merge branch 'master' into create_TC_CADMIN_1_22_test

j-ororke committed:
Merge branch 'master' into create_TC_CADMIN_1_22_test

j-ororke committed:
Update TC_CADMIN_1_22_24.py

Updated to include dependency location change for matter_testing support module

j-ororke committed:
Merge branch 'master' into create_TC_CADMIN_1_22_test

j-ororke committed:
Updated TC_CADMIN_1_22_24.py:
- Resolved issues with CI args in TC_CADMIN_1_22_24 test module
- Removed yaml calls for 1_21, 1_22, 1_23, and 1_24, also removed yaml call for 1_9 from manualTests.json

tehampson commented on src/python_testing/TC_CADMIN_1_22_24.py:
```diff
@@ -0,0 +1,171 @@
+
+        except Exception as e:
+            logging.exception('Error running OpenCommissioningWindow %s', e)
+            asserts.assert_true(False, 'Failed to open commissioning window')
```
```suggestion
            asserts.fail('Failed to open commissioning window')
```

j-ororke replied:
Thank you, the code has been updated to use this!

tehampson commented on src/python_testing/TC_CADMIN_1_22_24.py:
```diff
@@ -0,0 +1,171 @@
+
+        self.step(4)
+        window_status = await self.get_window_status()
+        if window_status != Clusters.AdministratorCommissioning.Enums.CommissioningWindowStatusEnum.kWindowNotOpen:
+            asserts.fail("Commissioning window is expected to be closed, but was found to be open")
```
I think this is the right code
```suggestion
        asserts.assert_equal(window_status, Clusters.AdministratorCommissioning.Enums.CommissioningWindowStatusEnum.kWindowNotOpen, "Commissioning window is expected to be closed, but was found to be open")
```

j-ororke replied:
Thank you, this has been tested and used in the current code!

tehampson commented on src/python_testing/TC_CADMIN_1_22_24.py:
```diff
@@ -0,0 +1,171 @@
+
+        self.step(6)
+        window_status2 = await self.get_window_status()
+        if window_status2 != Clusters.AdministratorCommissioning.Enums.CommissioningWindowStatusEnum.kWindowNotOpen:
```
Same issue here, just do `asserts.assert_equal`

tehampson commented on src/python_testing/TC_CADMIN_1_22_24.py:
```diff
@@ -0,0 +1,171 @@
+
+        self.step(4)
+        window_status = await self.get_window_status()
+        if window_status != Clusters.AdministratorCommissioning.Enums.CommissioningWindowStatusEnum.kWindowNotOpen:
```
same

j-ororke replied:
Thank you, I have updated this here now.

tehampson replied:
Please remove the if statement above and unnest the `asserts.assert_equal`

tehampson commented on src/python_testing/TC_CADMIN_1_22_24.py:
```diff
@@ -0,0 +1,171 @@
+
+        self.step(6)
+        window_status2 = await self.get_window_status()
+        if window_status2 != Clusters.AdministratorCommissioning.Enums.CommissioningWindowStatusEnum.kWindowNotOpen:
```
same

j-ororke replied:
Thank you, this has been updated

tehampson replied:
Please remove the if statement above and unnest the `asserts.assert_equal`

tehampson commented on src/python_testing/TC_CADMIN_1_22_24.py:
```diff
@@ -0,0 +1,171 @@
+        except ChipStackError as e:
+            # Since we provided 901 seconds as the timeout duration,
+            # we should not be able to open comm window as duration is too long.
+            asserts.assert_equal(e.err,  0x00000585,
```
For better readability I would suggest you do something like `_FAILURE_ERROR_CODE = 0x00000585` where `_FAILURE_ERROR_CODE` is a descriptive name about the errors like INVALID_TIMEOUT_ARGUMENT_ERROR, or whatever 0x00000585 means so that someone else can more easily see the error that you are expecting here

j-ororke replied:
Thank you,
That is a very good point, I have updated the comments to make those show what the error code is referring to

tehampson commented on src/python_testing/TC_CADMIN_1_22_24.py:
```diff
@@ -0,0 +1,171 @@
+                nodeid=self.dut_node_id, timeout=179, iteration=10000, discriminator=self.discriminator, option=1)
+
+        except ChipStackError as e:
+            # Since we provided 901 seconds as the timeout duration,
```
This comment seems like copy paste and should say 179

j-ororke replied:
Thank you, I have updated this comment as you are correct, I must have copied it from the above test and forgot to change the seconds used in this test.

tehampson commented on src/python_testing/TC_CADMIN_1_22_24.py:
```diff
@@ -0,0 +1,171 @@
+        except ChipStackError as e:
+            # Since we provided 901 seconds as the timeout duration,
+            # we should not be able to open comm window as duration is too long.
+            asserts.assert_equal(e.err,  0x00000585,
```
Again if you could update this to be more readable that would be great

j-ororke replied:
Thank you, I have updated the comments to be a bit cleaner here now

tehampson commented on src/python_testing/TC_CADMIN_1_22_24.py:
```diff
@@ -0,0 +1,171 @@
+        await self.th1.OpenCommissioningWindow(
+            nodeid=self.dut_node_id, timeout=180, iteration=10000, discriminator=self.discriminator, option=1)
+
+        self.step(3)
```
Step 3 for TC_CADMIN_1_24 in https://github.com/CHIP-Specifications/chip-test-plans/pull/4673/ says to wait 181 seconds for this to expire. It doesn't say to issue a revoke command. Should the testplan be updated or should you be waiting 181s?

j-ororke replied:
Hi Terence,
I have updated this to using the timeout of 180 seconds instead of the revoke commissioning command for step 3. Please note that where your comment is, is for TC_CADMIN_1_24 and the PR you mentioned was for TC_CADMIN_1_22. 
Note that there is an issue currently for when attempting to communicate with the endpoint after the timeout of the commissioning window, which I had filed an issue for already:
https://github.com/project-chip/connectedhomeip/issues/35983
We have a workaround already in place to get past that part until the issue is resolved, which you will notice in the latest change to the test module, which was is currently to just try again the get_window_status function.

j-ororke replied:
Ah that is a really good point, wish that it would allow for comments anywhere in the PR files, unfortunately, they block that. Thank you for the comment, I have updated as mentioned above to using the 180 second timeout during TC_CADMIN_1_24 and have continued with the revoke command for TC_CADMIN_1_22 as we just didn't want either the test harness or the testers having to wait the full 900 seconds before proceeding to the end of that test, which is due to the current issue #35983, we also have some other tests in CADMIN that perform the wait for timeout as well.

j-ororke committed:
Updating TC_CADMIN_1_22_24 test module:
- Resolving comments from tehampson as these were needed changes to be made
-- Substituted to using cleaner code for assert.asserts_equal instead of just assert.failure
-- Resolving copy paste error in TC_CADMIN_1_4 test
-- Made comments cleaner by rewording to include what the error code was referring too

j-ororke committed:
Update TC_CADMIN_1_22_24 test module:
- Changing method for test step 3 on TC_CADMIN_1_24 test now using timeout for 30 seconds instead of revoke commissioning

restyled-commits committed:
Restyled by autopep8

tehampson commented on src/python_testing/TC_CADMIN_1_22_24.py:
```diff
@@ -104,13 +106,14 @@ async def test_TC_CADMIN_1_22(self):
 
         self.step(6)
         window_status2 = await self.get_window_status()
         if window_status2 != Clusters.AdministratorCommissioning.Enums.CommissioningWindowStatusEnum.kWindowNotOpen:
```
This if statement can still be removed above

j-ororke replied:
Thank you, this has been completed!

tehampson commented on src/python_testing/TC_CADMIN_1_22_24.py:
```diff
@@ -104,13 +106,14 @@ async def test_TC_CADMIN_1_22(self):
             # Since we provided 901 seconds as the timeout duration,
             # we should not be able to open comm window as duration is too long.
+            # we are expected receive Failed to open commissioning window: IM Error 0x00000585: General error: 0x85 (INVALID_COMMAND)
             asserts.assert_equal(e.err,  0x00000585,
```
```suggestion
            _INVALID_COMMAND = 0x00000585
            asserts.assert_equal(e.err,  _INVALID_COMMAND,
```

Then you can replace all the other values of `0x00000585` with `_INVALID_COMMAND` and it makes the code more readable

j-ororke replied:
Thank you this does make the code much easier to read, i will update using your suggestion above!

tehampson reviewed (commented):
Overall looking better just a few more nits and I think we are good to go :)

j-ororke committed:
Update TC_CADMIN_1_22_24.py

- Unnesting asserts from if statements

j-ororke committed:
Update src/python_testing/TC_CADMIN_1_22_24.py

- Adding this suggested change from Terrence into the code, as it does make the code much cleaner

Co-authored-by: Terence Hampson <thampson@google.com>

j-ororke committed:
Update TC_CADMIN_1_22_24.py

-Updating to changing static value to being stored in variable for easier readability as Terrence has suggested, making the code easier to read.

restyled-commits committed:
Restyled by autopep8

j-ororke commented on src/python_testing/TC_CADMIN_1_22_24.py:
```diff
@@ -94,7 +95,8 @@ async def test_TC_CADMIN_1_22(self):
         self.step(4)
         window_status = await self.get_window_status()
         if window_status != Clusters.AdministratorCommissioning.Enums.CommissioningWindowStatusEnum.kWindowNotOpen:
```
Thank you, I have removed these if statements from the code

j-ororke committed:
Updated TC_CADMIN_1_22_24 test module:
- Removed unneccessary await from asserts

restyled-commits committed:
Restyled by autopep8

andy31415 commented on src/python_testing/TC_CADMIN_1_22_24.py:
```diff
@@ -0,0 +1,179 @@
+            nodeid=self.dut_node_id, timeout=180, iteration=10000, discriminator=self.discriminator, option=1)
+
+        self.step(3)
+        sleep(180)
```
yaml used to wait for 181 seconds instead of 180 for some reason.

Also this is absolutely terrible for CI. Could we make CI sleep less? Or maybe have an interface to skip slow in PR CI the same like we have for YAML: https://github.com/project-chip/connectedhomeip/blob/master/scripts/tests/chiptest/__init__.py#L98

j-ororke replied:
I apologize for this test having caused some issues in the CI currently due to the sleep time.
I will file a PR here soon to remove this test from running in the CI moving forward, sounds like this test is not needed to be run in the CI.

j-ororke replied:
Hi @andy31415, 
I noticed that there is some new code in src/python_testing/test_metadata.yaml in the master branch.
In the slow_tests section of that yaml file there appears to be CADMIN_1_9, CADMIN_1_19, and CADMIN_1_22_24 tests with duration of test runs added currently. 
Does this resolve the issue with the CI and the test from this above PR? Can this comment be resolved now?

Should the CADMIN_1_22_24 python test module also be added to the "not_automated" section with a caveat that it is causing issues due to long run time in order for it to not be selected in the execute_python_tests.py script during test runs in CI by the test runners?
